recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,259 @@
+  moduleSecret string                  // SecretStoreId associated with git private terraform repo source
+  providerSecrets map[string]map[string]SecretReference  // map[providerName or provider.aliasName][secret_name]SecretReference
+  envSecrets map[string]SecretReference // map[environmentvariablename]SecretReference 
+}
+
+type RecipeSecrets struct {
+  moduleSecrets v20231001preview.SecretStoresClientListSecretsResponse
+  providerSecrets map[string]map[string]string   // map[providerName/provider.aliasName][secret_name]value
+  envSecrets map[string]string                   // map[environmentvariablename]value 
+}
+
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface to now return RecipeSecretIDs
+type DriverWithSecrets interface {
+	// Driver is an interface to implement recipe deployment and recipe resources deletion.
+	Driver
+
+	// FindSecretIDs gets the secret store resource IDs for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
+}
+
+LoadSecrets in SecretsLoader interface will be updated to LoadRecipeSecrets(), It will internally call functions which will populate resolved secrets based on the Ids in RecipeSecretIDs{}. The result is combined as RecipeSecrets{} which is passed as BaseOptions.Secrets to the Driver in Execute() call.
+
+LoadRecipeSecrets(RecipeSecretIDs, ...)(RecipeSecrets, error){
+  ...
+  loadModuleSecrets()
+  loadProviderSecrets()
+  loadEnvSecrets()
+}
+
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+	// Configuration is the configuration for the recipe.
+	Configuration recipes.Configuration
+
+	// Recipe is the recipe metadata.
+	Recipe recipes.ResourceMetadata
+
+	// Definition is the environment definition for the recipe.
+	Definition recipes.EnvironmentDefinition
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets **RecipeSecrets**
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
+
+## Test plan
+
+Testing will include unit tests on all new functions created and a
+functional Test will be created to test this functionality e2e.
+
+## Security
+
+Secret data will be held in memory and be passed to the Driver which will inject these values into the Terraform configuration in working directory for Terraform. This directory is created when a Terraform recipe is set to be deployed and deleted once deployment of the Terraform recipe is completed..
+
+Secret data will not be persisted and secret data will not be logged by Radius.
+Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.
```

Comment:
> Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.

Is this relevant to this work though? We aren't really working with variables here.

Reply:
This was a general statement but I had a usecase in mind. i will try a test with an example and write back.

Reply:
Will wait for this to be updated. It's unclear how sensitive flag in the recipe impacts what Radius does here.

Reply:
I tested this and removed this sentence. yes, Radius can work with variables marked sensitive but there is no impact to what Radius does.